main.go

```diff
@@ -1,99 +1,61 @@
+	master     string
+	kubeconfig string
+	exitCode   int
+)
 
-	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+func main() {
+	// setup args
+	flag.BoolVar(&debug, "debug", false, "Print debug info")
 	flag.StringVar(&master, "master", "", "master url")
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
 	flag.Parse()
 
-	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
-	if err != nil {
-		return err
+	// setup logging
+	if debug {
+		log.SetLevel(log.DebugLevel)
 	}
 
-	clientset, err := routeclientset.NewForConfig(config)
+	// build config
+	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
 	if err != nil {
-		return err
+		log.Fatal(err)
 	}
 
+	// setup informer stop channel
 	stop := make(chan struct{})
-	return routesController(stop, clientset)
-}
+	defer close(stop)
 
-func routesController(stopCh chan struct{}, clientset *routeclientset.Clientset) error {
-	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
-	informer := factory.Route().V1().Routes().Informer()
-	defer close(stopCh)
-	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			route := obj.(*v1.Route)
-			log.Infof("added: %s %s", route.GetName(), route.Spec.Host)
-			if err := expose(route.Spec.Host); err != nil {
-				log.Error(err)
-			}
-		},
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			old := oldObj.(*v1.Route)
-			route := newObj.(*v1.Route)
-			if old.Spec.Host != route.Spec.Host {
-				log.Infof("updated: %s (%s -> %s)", route.GetName(), old.Spec.Host, route.Spec.Host)
-				if err := unexpose(old.Spec.Host); err != nil {
-					log.Error(err)
-				}
-				if err := expose(route.Spec.Host); err != nil {
-					log.Error(err)
-				}
-			}
-		},
-		DeleteFunc: func(obj interface{}) {
-			route := obj.(*v1.Route)
-			log.Infof("deleted: %s %s", route.GetName(), route.Spec.Host)
-			if err := unexpose(route.Spec.Host); err != nil {
-				log.Error(err)
-			}
-		},
-	})
-	informer.Run(stopCh)
-	return nil
-}
-
-func expose(host string) error {
-	bin, err := json.Marshal([]string{host})
+	// run node port handler
+	nodePortClientSet, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return err
+		log.Fatal(err)
 	}
-	_, err = http.Post("http://gateway/hosts/add", "application/json", bytes.NewReader(bin))
-	return err
-}
+	nodePortHandler := nodeporthandler.NodePortHandler(nodePortClientSet)
+	go func() {
+		nodePortHandler.Run(stop)
+	}()
 
-func unexpose(host string) error {
-	bin, err := json.Marshal([]string{host})
+	// run routes handler
+	routesClientSet, err := routeclientset.NewForConfig(config)
 	if err != nil {
-		return err
+		log.Fatal(err)
 	}
-	_, err = http.Post("http://gateway/hosts/remove", "application/json", bytes.NewReader(bin))
-	return err
+	routePortHandler := routeshandler.RoutesHandler(routesClientSet)
+	go func() {
+		routePortHandler.Run(stop)
+	}()
```

Comment:
With both `nodePortHandler.Run()` and `routePortHandler.Run()` in their own go routines, a loop is needed to prevent `main()` from exiting right away. A `for { time.Sleep(30*time.Second) }` should do the trick, though you could also catch signals to exit cleanly:
```
	signalCh := make(chan os.Signal, 1)
	signal.Notify(signalCh, syscall.SIGTERM)
	select {
		case <-signalCh:
			return
	}
```